src/wallet/coin_selection.rs

```diff
@@ -171,6 +145,178 @@ impl CoinSelectionResult {
+/// - `optional_utxos`: the remaining available utxos to satisfy `target_amount` with their
+///                     weight cost
+/// - `fee_rate`: fee rate to use
+/// - `target_amount`: the outgoing amount in satoshis and the fees already
+///                    accumulated from added outputs and transaction’s header.
+/// - `weighted_drain_script`: the drain script used to create the change output associated with
+///                            the max satisfaction weight needed to spend it
+pub fn get_selection<D: Database, Cs: CoinSelectionAlgorithm<D>>(
+    algorithm: Cs,
+    database: &D,
+    required_utxos: Vec<WeightedUtxo>,
+    optional_utxos: Vec<WeightedUtxo>,
+    fee_rate: FeeRate,
+    target_amount: u64,
+    weighted_drain_script: &WeightedScript,
+) -> Result<CoinSelectionResult, Error> {
+    // ####################################################################
+    // ######################### PREPROCESSING ############################
+    // ####################################################################
+
+    // Mapping every (UTXO, usize) to an output group
+    let mut required_utxos: Vec<OutputGroup> = required_utxos
+        .into_iter()
+        .map(|u| OutputGroup::new(u, fee_rate))
+        .collect();
+
+    // Mapping every (UTXO, usize) to an output group.
+    let optional_utxos: Vec<OutputGroup> = optional_utxos
+        .into_iter()
+        .map(|u| OutputGroup::new(u, fee_rate))
+        .filter(|u| u.effective_value.is_positive())
+        .collect();
+
+    let req_values = required_utxos.iter().fold((0, 0), |(eff_value, fees), x| {
+        (eff_value + x.effective_value, fees + x.fee)
+    });
+
+    let required_effective_value = req_values.0;
+    let required_fees = req_values.1;
+
+    let opt_values = optional_utxos.iter().fold((0, 0), |(eff_value, fees), x| {
+        (eff_value + x.effective_value, fees + x.fee)
+    });
+
+    let optional_effective_value = opt_values.0;
+    let optional_fees = opt_values.1;
+
+    // `required_effective_value` and `optional_effective_value` are both the sum of *effective_values* of
+    // the UTXOs. For the optional UTXOs (optional_effective_value) we filter out UTXOs with
+    // negative effective value, so it will always be positive.
+    //
+    // Since we are required to spend the required UTXOs (required_effective_value) we have to consider
+    // all their effective values, even when negative, which means that required_effective_value could
+    // be negative as well.
+    //
+    // If the sum of required_effective_value and optional_effective_value is negative or lower than our target,
+    // we can immediately exit with an error, as it's guaranteed we will never find a solution
+    // if we actually run the coin selection algorithm
+    let total_value: Result<u64, _> =
+        (optional_effective_value + required_effective_value).try_into();
+    match total_value {
+        Ok(v) if v >= target_amount => {}
```

Comment:
If `total_value == target_amount`, we could just skip coin selection right?

Reply:
Yes!